### DISCUSSION TITLE
FabricJS object.set() gives not a function error when loading canvas from json

### DISCUSSION THREAD

**Discussion by SirPm** [fabricjs/fabric.js#7545] [Q&A]
After I perform actions with fabric js, I convert the content of the canvas to json and save it to localstorage by calling saveAsJson() and when the component mounts because I'm using react, I call the loadContentBack() function which loads back perfectly. But when I start to perform actions on the object I get the error, object.set is not a function, and this error only comes up when I load the canvas content from json.

`function saveAsJson() {
        if (canvas) {
            localStorage.setItem("canvasJson", JSON.stringify(canvas.toJSON(["id"])));
        }
}
  
 function loadContentBack() {
  const jsonData = JSON.parse(localStorage.getItem("canvasJson") ?? JSON.stringify(canvas.toJSON(["id"])));
  
  canvas.loadFromJSON(jsonData, canvas.renderAll.bind(canvas));
 }`

**Comment by asturur** [comment]
loadFromJSON takes a callback and sometimes is asyncronous.
Are you waiting for it to finish load before executing actions?

**Reply by SirPm** [reply]
Yes I think so, I'm doing something like this now

canvas.loadFromJSON(jsonData, canvas.renderAll.bind(canvas), function (o: any, object: any) {
	console.log(object, "the object now you can perform actions");
});

I don't know how to tell if it has finished loading, except through this callback function. So I only added two components. Saved it and waited for the callback function to log before performing any operation but still got the same error.

I honestly don't know what I'm doing wrong at this point

**Reply by asturur** [reply]
So the issue is that `canvas.renderAll.bind(canvas)` is your callback.

```js
canvas.loadFromJSON(jsonData, function() {
  // it finished loading!
  // do something!
  canvas.requestRenderAll();
}, function (o: any, object: any) {
  // this is a reviver function.
  // o is the fabric object instance
  // object is its serialized counterpart from which the object was revived
});

```

**Reply by SirPm** [reply]
Alright it's still not working, the components load properly i can move them but I cannot change any of the properties of the component like color etc.

I checked the documentation again, I want to understand what the reviver function does but can't see a clear explanation there.
Am I to still do something else after the components have been loaded back before I can perform operations on them?

**Reply by asturur** [reply]
i don't think the reviver function is what you need.
can you show a larger example of what are you doing with actual JSDON data?

**Reply by SirPm** [reply]
For now I'm not doing anything with the json data I'm just loading it back to the canvas.

But I think I just found out what the issue is. In a section of my code I have a place where I listen for the "selection:updated" event, when I do this I update the current component in view.
But in the object from the callback the width and height are not showing the updated one. So I recalculate it and create an object by spreading the previous properties and add the recalculated width and height. 

```js
canvas.on("selection:updated", function (obj: any) {
				if (obj.target && obj.target.width && obj.target.height && obj.target.scaleY && obj.target.scaleX) {
					console.log(obj, "updated");
					const newWidth = Math.round(obj.target.width * obj.target.scaleX);
					const newHeight = Math.round(obj.target.height * obj.target.scaleY);

					console.log(newWidth);
					console.log(newHeight);
					const updatedObject = { ...obj.target, width: newWidth, height: newHeight } as fabric.Object;
setActiveObject(updatedObject);
				}
			});
```
But in creating the object like this, it no longer behaves like the normal fabric object, that's why I was getting the object.set is not a function.

So now I'm looking for a way to convert the created object to a fabric object.
I tried doing type casting, but it didn't work. And I couldn't create a new instance using new fabric.Object()  because this will create a new object which doesn't work as expected

Right now I'm looking for a way to convert the object I created to a fabric object without creating a new instance

**Comment by MunizErthal** [comment]
I have the same problema...
When I use `object.set({opacity: 1});` get the same error.

But I discovered something today.
If I select the object with event:

`this.canvas.on('mouse:down',  e => { this.clickDown(e)  });`

And use `e.target.set({opacity: 1});` set works

**Reply by melchiar** [reply]
Hi, can you add a fiddle or codepen showing the issue?
